lib/src/test/java/org/kpipe/consumer/OffsetManagerTest.java

```diff
@@ -0,0 +1,798 @@
+package org.kpipe.consumer;
+
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.ArgumentMatchers.*;
+import static org.mockito.Mockito.*;
+
+import java.time.Duration;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.common.TopicPartition;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.kpipe.consumer.enums.ConsumerCommand;
+import org.mockito.*;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
+class OffsetManagerTest {
```

Comment:
i dont see any tests that test correctness of highestProcessed offset commits - when there are no pending offsets left...
Changing one of the tests with gaps - to have no gaps - so everything marked as processed - actually fails the test - as offset committed is too low (or at least nextOffsetToCommit is too low)...

Reply:
overall IMHO - i would suggest to use awaitility with Assertions / AssertJ asserts for eventual assertions - that way proper asserts can be used that log both expected and actual values on failure etc...
it is personal preference again of course - but the AssertJ assertions are nice and fluent... while avaitility adds the eventual assertion handling quite nicely.

Reply:
another IMHO - it maybe worth adding couple of integration tests - its easy to spin up docker container for Kafka - and tests can be easily isolated by topic name - so only need to spin up / down Kafka once per test set... 
But they would flesh out any inconsistencies with mocks / overloads quite fast / easily...

Reply:
> overall IMHO - i would suggest to use awaitility with Assertions / AssertJ asserts for eventual assertions - that way proper asserts can be used that log both expected and actual values on failure etc...
it is personal preference again of course - but the AssertJ assertions are nice and fluent... while avaitility adds the eventual assertion handling quite nicely.

Yeah I agreed 100%. I logged while ago #50

Reply:
> i dont see any tests that test correctness of highestProcessed offset commits - when there are no pending offsets left...
Changing one of the tests with gaps - to have no gaps - so everything marked as processed - actually fails the test - as offset committed is too low (or at least nextOffsetToCommit is too low)...

Can you provide an scenario for this test case?